src/main/java/io/vertx/core/shareddata/impl/LocalAsyncLocks.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2011-2018 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.shareddata.impl;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.VertxException;
+import io.vertx.core.shareddata.Lock;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @author Thomas Segismont
+ */
+public class LocalAsyncLocks {
+
+  private enum Status {WAITING, ACQUIRED, TIMED_OUT}
+
+  private class LockWaiter {
+
+    final Context context;
+    final String lockName;
+    final Handler<AsyncResult<Lock>> handler;
+    final AtomicReference<Status> status;
+    final Long timerId;
+
+    LockWaiter(Context context, String lockName, long timeout, Handler<AsyncResult<Lock>> handler) {
+      this.context = context;
+      this.lockName = lockName;
+      this.handler = handler;
+      status = new AtomicReference<>(Status.WAITING);
+      timerId = timeout != Long.MAX_VALUE ? context.owner().setTimer(timeout, tid -> timeout()) : null;
+    }
+
+    boolean isWaiting() {
+      return status.get() == Status.WAITING;
+    }
+
+    void timeout() {
+      if (status.compareAndSet(Status.WAITING, Status.TIMED_OUT)) {
+        handler.handle(Future.failedFuture(new VertxException("Timed out waiting to get lock")));
+      }
+    }
+
+    void acquireLock() {
+      if (status.compareAndSet(Status.WAITING, Status.ACQUIRED)) {
+        if (timerId != null) {
+          context.owner().cancelTimer(timerId);
+        }
+        context.runOnContext(v -> handler.handle(Future.succeededFuture(new AsyncLock(lockName))));
+      } else {
+        context.runOnContext(v -> nextWaiter(lockName));
+      }
+    }
+  }
+
+  private class AsyncLock implements Lock {
+    final String lockName;
+    final AtomicBoolean invoked = new AtomicBoolean();
+
+    AsyncLock(String lockName) {
+      this.lockName = lockName;
+    }
+
+    @Override
+    public void release() {
+      if (invoked.compareAndSet(false, true)) {
```

Comment:
`java.util.concurrent.locks.Lock` throws `java.lang.IllegalMonitorStateException` when you release an not owned lock, perhaps we should do the same here ?

Reply:
I don't believe we should. But this PR is supposed to fix #2484 , so I'd prefer we stick to that.

Reply:
fine